src/Evolution/Systems/GeneralizedHarmonic/BoundaryConditionsImpl.hpp

```diff
@@ -0,0 +1,1447 @@
+            spacetime_unit_normal_vector,
+        const tnsr::aa<DataVector, VolumeDim, Frame::Inertial>& projection_ab,
+        const tnsr::Ab<DataVector, VolumeDim, Frame::Inertial>& projection_Ab,
+        const tnsr::AA<DataVector, VolumeDim, Frame::Inertial>& projection_AB,
+        const tnsr::II<DataVector, VolumeDim, Frame::Inertial>&
+            inverse_spatial_metric,
+        const tnsr::ii<DataVector, VolumeDim, Frame::Inertial>&
+            extrinsic_curvature,
+        const tnsr::aa<DataVector, VolumeDim, Frame::Inertial>&
+            spacetime_metric,
+        const tnsr::AA<DataVector, VolumeDim, Frame::Inertial>&
+            inverse_spacetime_metric,
+        const typename Tags::ThreeIndexConstraint<
+            VolumeDim, Frame::Inertial>::type& three_index_constraint,
+        const tnsr::aa<DataVector, VolumeDim, Frame::Inertial>&
+            char_projected_rhs_dt_u_minus,
+        const tnsr::iaa<DataVector, VolumeDim, Frame::Inertial>& phi,
+        const tnsr::ijaa<DataVector, VolumeDim, Frame::Inertial>& d_phi,
+        const tnsr::iaa<DataVector, VolumeDim, Frame::Inertial>& d_pi,
+        const std::array<DataVector, 4>& char_speeds) noexcept {
+  //-------------------------------------------------------------------
+  // Add in physical boundary condition
+  //-------------------------------------------------------------------
+  ASSERT(get_size(get<0, 0>(*bc_dt_u_minus)) == get_size(get(gamma2)),
+         "Size of input variables and temporary memory do not match.");
+  // hard-coded value from SpEC Bbh input file Mu = MuPhys = 0
+  const double mMuPhys = 0.;
+  const bool mAdjustPhysUsingC4 = true;
+  const bool mGamma2InPhysBc = true;
+
+  // In what follows, we use the notation of Kidder, Scheel & Teukolsky
+  // (2001) https://arxiv.org/pdf/gr-qc/0105031.pdf. We will refer to this
+  // article as KST henceforth, and use the abbreviation in variable names
+  // to disambiguate their origin.
+  auto U8p = make_with_value<tnsr::ii<DataVector, VolumeDim, Frame::Inertial>>(
+      unit_interface_normal_vector, 0.);
+  auto U8m = make_with_value<tnsr::ii<DataVector, VolumeDim, Frame::Inertial>>(
+      unit_interface_normal_vector, 0.);
+  {
+    // D_(k,i,j) = (1/2) \partial_k g_(ij) and its derivative
+    tnsr::ijj<DataVector, VolumeDim, Frame::Inertial> kst_D(
+        get_size(get(gamma2)));
+    tnsr::ijkk<DataVector, VolumeDim, Frame::Inertial> d_kst_D(
+        get_size(get(gamma2)));
+    for (size_t i = 0; i < VolumeDim; ++i) {
+      for (size_t j = i; j < VolumeDim; ++j) {
+        for (size_t k = 0; k < VolumeDim; ++k) {
+          kst_D.get(k, i, j) = 0.5 * phi.get(k, i + 1, j + 1);
+          for (size_t l = 0; l < VolumeDim; ++l) {
+            d_kst_D.get(l, k, i, j) = 0.5 * d_phi.get(l, k, i + 1, j + 1);
+          }
+        }
+      }
+    }
+
+    // ComputeRicciFromD(kst_D, d_kst_D, Invg, Ricci3);
+    auto ricci_3 = GeneralizedHarmonic::spatial_ricci_tensor_from_KST_vars(
```

Comment:
why can't you use the existing pointwise functions to get the 3-ricci and christoffel second kind?

Reply:
Changed these to use existing christoffel functions, but I had to add an overload of the `ricci_tensor` function anyway as the existing function (there's only one) for which need as input quantities that haven't been computed here yet.